### PR TITLE
Feature/query string filter

### PR DESCRIPTION
**Pull request by eladex**
* generic filter query for mongoDB
* generic filter and search query for elasticSearch
* transform query middleware:
filter query fields
transform to camel case
apply query defaults
swap query aliases with values
* use enum for `STATUS` field

### CODE REVIEW

File: src/config/db-enums.ts [hunk 1]
```diff
@@ -25,3 +23,7 @@ export const DOMAIN_MAP = [
 export const SERVICE_TYPE = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H'];
 
 export const DATA_SOURCE = ['dataSource1', 'dataSource2','aka', 'es_name', 'ads_name', 'adNN_name', 'nvSQL_name', 'lmn_name', 'mdn_name', 'mm_name', 'city_name'];
+
+export enum STATUS { ACTIVE = 'active', INACTIVE = 'inactive', INCOMPLETE = 'incomplete' }
+
+export const ALL_STATUS = Object.keys(STATUS).map(k => STATUS[k]);
```
Comment (eladex) [inline src/config/db-enums.ts]: maybe transfer to utils

File: src/helpers/queryTransform.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  queryValuesAliases: ObjectValueMap = {}) {
+  // rename query params
+  const renamed = transformKeys(originalQuery, queryParamsRenameMap);
+  // filter params
+  const filtered = filterQueryParams ? filterObjectByKeys(renamed, filterQueryParams) : renamed;
+  // apply defaults
+  const tQuery = { ...queryDefaults, ...filtered };
+  // apply values aliases and return 
+  return transformValues(tQuery, queryValuesAliases);
+}
```
Comment (eladex) [inline src/helpers/queryTransform.ts]: add docs

File: src/helpers/queryTransform.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+function transformQuery(
+  originalQuery: object,
+  queryParamsRenameMap: KeyMap = {},
+  filterQueryParams: string[] = null,
+  queryDefaults: ValueMap = {}, 
+  queryValuesAliases: ObjectValueMap = {}) {
+  // rename query params
+  const renamed = transformKeys(originalQuery, queryParamsRenameMap);
+  // filter params
+  const filtered = filterQueryParams ? filterObjectByKeys(renamed, filterQueryParams) : renamed;
```
Comment (eladex) [inline src/helpers/queryTransform.ts]: switch order and add  "case insensitive" flag in `filterObjectByKeys`

File: src/utils.ts [hunk 4]
```diff
@@ -4,6 +4,20 @@ import { DOMAIN_MAP } from './config/db-enums';
 export const domainMap : Map<string, string> = new Map<string, string>(JSON.parse(JSON.stringify(DOMAIN_MAP)));
 export const DomainSeperator = '@';
 
+export type BasicType = boolean | string | number | any[];
+
+export interface ValueMap {
+  [key: string]: BasicType;
```
Comment (eladex) [inline src/utils.ts]: `  [key: string]: BasicType | BasicType[]`

File: src/helpers/repository.ts [hunk 5]
```diff
@@ -17,6 +17,18 @@ export interface ICollection<T> {
...
+  const cond = {};
+  for (const field of Object.keys(queryObj)) {
+    const value = queryObj[field];
+    if (Array.isArray(value)) {
+      cond[field] = { $in: value };
+    }
+    cond[field] = value;
+  }
+  return cond;
+}
```
Comment (eladex) [inline src/helpers/repository.ts]: add docs

File: src/helpers/repository.ts [hunk 6]
```diff
@@ -17,6 +17,18 @@ export interface ICollection<T> {
   totalCount: number;
 }
 
+function queryParser(queryObj: object): any {
+  const cond = {};
+  for (const field of Object.keys(queryObj)) {
```
Comment (eladex) [inline src/helpers/repository.ts]: use `Object.entries()`

File: src/person/person.route.ts [hunk 7]
```diff
@@ -6,23 +6,25 @@ import { AuthMiddleware } from '../middlewares/auth.middleware';
 import { Person } from './person.controller';
 // import { IPerson, EDITABLE_FIELDS, PERSON_FIELDS } from './person.interface';
 import { validatorMiddleware, RouteParamsValidate as Vld } from '../helpers/route.validator';
-import { atCreateFieldCheck, atUpdateFieldCheck, atSearchFieldCheck } from './person.route.validator';
+import { atCreateFieldCheck, atUpdateFieldCheck, atSearchFieldCheck, queryAllowedFields } from './person.route.validator';
```
Comment (eladex) [inline src/person/person.route.ts]: delete `atSearchFieldCheck` and  `queryAllowedFields `

File: src/person/person.queryMiddleware.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+};
+
+const queryValuesAliases = config.queries.queryAliases.persons;
+
+export const queryAllowedFields = ['currentUnit', 'domainUsers.dataSource', 'entityType', 
+  'job', 'rank', 'responsibility', 'serviceType', 'status'];
+
+export const serachAllowedFields = queryAllowedFields.concat(['fullName']); 
+
+export const queryMiddleware = makeMiddleware({
```
Comment (eladex) [inline src/person/person.queryMiddleware.ts]: maybe rename to `filterQueryMiddleware`

File: src/person/person.queryMiddleware.ts [hunk 9]
```diff
@@ -0,0 +1,49 @@
+// import { KeyMap } from '../utils';
```
Comment (eladex) [inline src/person/person.queryMiddleware.ts]: delete

File: src/person/person.route.validator.ts [hunk 10]
```diff
@@ -5,10 +5,15 @@ import { RouteParamsValidate as RPV } from '../helpers/route.validator';
...
-  'responsibilityLocation', 'serviceType','status'];
+  'responsibilityLocation', 'serviceType', 'status'];
 
 export const updateAllowedFields = _.without(createAllowedFileds, 'directGroup', 'identityCard', 'personalNumber');
 
+export const queryAllowedFields = ['currentUnit', 'domainUsers', 'domainUsers.dataSource', 'entityType', 
+  'firstName','job', 'lastName', 'rank', 'responsibility', 'serviceType', 'status'];
+
+export const serachAllowedFields = queryAllowedFields.concat(['fullName']); 
+
```
Comment (eladex) [inline src/person/person.route.validator.ts]: delete

File: src/person/person.route.ts [hunk 11]
```diff
@@ -6,23 +6,25 @@ import { AuthMiddleware } from '../middlewares/auth.middleware';
...
 
 persons.use('/', AuthMiddleware.verifyToken, PermissionMiddleware.hasBasicPermission);
 
-persons.get('/', ch(Person.getPersons, (req: Request) => [req.query]));
+persons.get('/', queryMiddleware, ch(Person.getPersons, (req: Request) => [req.query]));
 
-persons.get('/search', validatorMiddleware(atSearchFieldCheck, null, 'query'),
-  ch(Person.autocomplete, (req: Request) => [req.query.fullname]));
+persons.get('/search', searchMiddleware,
+  // validatorMiddleware(atSearchFieldCheck, null, 'query'),
```
Comment (eladex) [inline src/person/person.route.ts]: delete

File: src/person/person.route.ts [hunk 12]
```diff
@@ -6,23 +6,25 @@ import { AuthMiddleware } from '../middlewares/auth.middleware';
...
+  ch(Person.searchPersons, (req: Request) => [req.query]));
 
 persons.get('/getUpdated/:from', validatorMiddleware(Vld.dateOrInt, ['from'], 'params'),
```
Comment (eladex) [inline src/person/person.route.ts]: add `queryMiddleware` !!!

File: src/person/person.spec.ts [hunk 13]
```diff
@@ -220,6 +220,21 @@ describe('Persons', () => {
       persons[1].should.have.property('personalNumber', '3456712');
       persons[2].should.have.property('personalNumber', '4567123');
     });
+    it('should get a person from a given date and specific entity type', async () => {
```
Comment (eladex) [inline src/person/person.spec.ts]: add tests that filters with multiple fields

Comment (eladex) [reply src/person/person.spec.ts]: there is already a test for that

File: src/config/db-enums.ts (hunk 1 above)
Comment (eladex) [reply src/config/db-enums.ts]: transferred

File: src/utils.ts [hunk 14]
```diff
@@ -1,13 +1,15 @@
 import { ValidatorObj } from './types/validation';
-import { DOMAIN_MAP } from './config/db-enums';
+import { DOMAIN_MAP, STATUS } from './config/db-enums';
+import * as _ from 'lodash';
```
Comment (eladex) [inline src/utils.ts]: delete import

File: openapi.yaml [hunk 15]
```diff
@@ -22,9 +22,19 @@ paths:
     get:
       tags:
         - Person
-      summary: Returns a list of all persons
+      summary: Gets persons with optional filters
```
Comment (eladex) [inline openapi.yaml]: Add filter parameters to /person/updatedFrom

**Comment by DavidHeymann** [comment]
well done

**Pull request merged**